### PR TITLE
[Feature]Support desc table of external catalog

### PR DESCRIPTION
**Pull request by leoyy0316**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #https://github.com/StarRocks/starrocks/issues/8761 https://github.com/StarRocks/starrocks/issues/7285

## Problem Summary(Required) ：
In order to support desc table, first need to implement 'show proc /catalog' .
**The implementation "'show proc catalog'" logic is as follows：**
1. 'show proc '/catalog' contains "default_catalog" and other external catalog;
    MySQL [(none)]> show proc '/catalog';
+-----------------+----------+------------------+
| Catalog         | Type     | Comment          |
+-----------------+----------+------------------+
| hive_catalog    | hive     | NULL             |
| default_catalog | Internal | Internal Catalog |
+-----------------+----------+------------------+

2.show proc '/catalog/hive_catalog'
3.show proc '/catalog/hive_catalog/temp_db'
4.show proc '/catalog/hive_catalog/temp_db/test_leo'
5.show proc '/catalog/hive_catalog/temp_db/test_leo/schema'
MySQL [(none)]> show proc '/catalog/hive_catalog/temp_db/test_leo/schema';
+-------+----------------+------+-------+---------+-------+
| Field | Type           | Null | Key   | Default | Extra |
+-------+----------------+------+-------+---------+-------+
| id    | INT            | Yes  | false | NULL    |       |
| name  | VARCHAR(65533) | Yes  | false | NULL    |       |
+-------+----------------+------+-------+---------+-------+

show proc '/catalog/default_catalog ' logic same as show '/dbs'

desc  external table implement through the catalog proc system. The returned result is the same with  "show proc '/catalog/hive_catalog/temp_db/test_leo/schema'"

MySQL [(none)]> desc hive_catalog.temp_db.test_leo;
+-------+----------------+------+-------+---------+-------+
| Field | Type           | Null | Key   | Default | Extra |
+-------+----------------+------+-------+---------+-------+
| id    | INT            | Yes  | false | NULL    |       |
| name  | VARCHAR(65533) | Yes  | false | NULL    |       |
+-------+----------------+------+-------+---------+-------+

### CODE REVIEW

**Comment by leoyy0316** [comment]
run starrocks_fe_unittest

**Comment by leoyy0316** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 26 / 180 (14.44%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Table.java|0|1|00.00%|[206]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTablesProcDir.java|0|43|00.00%|[20, 21, 22, 27, 28, 29, 30, 34, 39, 40, 42, 43, 44, 45, 47, 48, 50, 51, 52, 55, 57, 62, 63, 64, 66, 67, 68, 69, 71, 72, 73, 74, 75, 77, 78, 79, 80, 81, 82, 83, 84, 86, 87]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|0|10|00.00%|[123, 124, 127, 128, 129, 130, 131, 132, 133, 134]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalSchemaProcNode.java|0|24|00.00%|[17, 18, 19, 20, 24, 27, 28, 29, 33, 35, 36, 38, 39, 40, 41, 44, 45, 46, 47, 48, 49, 52, 53, 54]|
|:large_blue_circle:|com/starrocks/analysis/DescribeStmt.java|0|2|00.00%|[132, 133]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTableProcDir.java|0|23|00.00%|[15, 16, 17, 21, 22, 23, 24, 28, 29, 30, 34, 36, 37, 38, 39, 40, 45, 50, 52, 53, 57, 58, 60]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalDbsProcDir.java|0|36|00.00%|[16, 17, 18, 21, 22, 23, 27, 28, 29, 30, 32, 34, 35, 36, 37, 38, 40, 44, 45, 46, 47, 48, 50, 51, 53, 54, 56, 58, 59, 61, 62, 67, 72, 73, 74, 76]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|11|21|52.38%|[197, 203, 209, 249, 356, 357, 359, 360, 361, 362]|
|:large_blue_circle:|com/starrocks/common/proc/CatalogProcNode.java|14|19|73.68%|[22, 27, 28, 30, 41]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|1|1|100.00%|[]|

**Comment by leoyy0316** [comment]
run starrocks_fe_unittest

**Comment by leoyy0316** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 130 / 180 (72.22%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/DescribeStmt.java|0|2|00.00%|[132, 133]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|11|21|52.38%|[197, 203, 209, 249, 356, 357, 359, 360, 361, 362]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTablesProcDir.java|26|43|60.47%|[34, 39, 40, 42, 43, 44, 45, 47, 48, 50, 51, 52, 55, 57, 67, 68, 75]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTableProcDir.java|16|23|69.57%|[45, 50, 52, 53, 57, 58, 60]|
|:large_blue_circle:|com/starrocks/common/proc/CatalogProcNode.java|14|19|73.68%|[22, 27, 28, 30, 41]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalDbsProcDir.java|27|36|75.00%|[35, 36, 40, 48, 67, 72, 73, 74, 76]|
|:large_blue_circle:|com/starrocks/catalog/Table.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalSchemaProcNode.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|1|1|100.00%|[]|

**Comment by leoyy0316** [comment]
run starrocks_fe_unittest

**Comment by leoyy0316** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 130 / 180 (72.22%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/DescribeStmt.java|0|2|00.00%|[132, 133]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|11|21|52.38%|[197, 203, 209, 249, 356, 357, 359, 360, 361, 362]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTablesProcDir.java|26|43|60.47%|[34, 39, 40, 42, 43, 44, 45, 47, 48, 50, 51, 52, 55, 57, 67, 68, 75]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTableProcDir.java|16|23|69.57%|[45, 50, 52, 53, 57, 58, 60]|
|:large_blue_circle:|com/starrocks/common/proc/CatalogProcNode.java|14|19|73.68%|[22, 27, 28, 30, 41]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalDbsProcDir.java|27|36|75.00%|[35, 36, 40, 48, 67, 72, 73, 74, 76]|
|:large_blue_circle:|com/starrocks/catalog/Table.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalSchemaProcNode.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/analysis/DescribeStmt.java [hunk 1]
```diff
@@ -80,6 +80,8 @@ public class DescribeStmt extends ShowStmt {
     private boolean isOlapTable;
     private boolean isMaterializedView;
 
+    private boolean isExternalTable;
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/analysis/DescribeStmt.java:83]: Is this variable useful?

File: fe/fe-core/src/main/java/com/starrocks/server/CatalogMgr.java [hunk 2]
```diff
@@ -199,6 +197,10 @@ private static class SerializeData {
 
     }
 
+    public ConcurrentHashMap<String, Catalog> getCatalogs() {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/server/CatalogMgr.java:200]: It destroys encapsulation, Don't use getCatalogs()

File: fe/fe-core/src/main/java/com/starrocks/server/CatalogMgr.java [hunk 3]
```diff
@@ -222,20 +224,4 @@ private void writeLock() {
     private void writeUnLock() {
         this.catalogLock.writeLock().unlock();
     }
-
-    public class CatalogProcNode implements ProcNodeInterface {
-        @Override
-        public ProcResult fetchResult() {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/server/CatalogMgr.java:228]: you can define your method in here.

File: fe/fe-core/src/test/java/com/starrocks/common/proc/ExternalProcDirTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    @Test
+    public void testExternalDbsFetchResultNormal() {
+        ExternalDbsProcDir dir = new ExternalDbsProcDir("hive_catalog");
+        ProcResult result = dir.fetchResult();
+        Assert.assertNotNull(result);
+        Assert.assertTrue(result instanceof BaseProcResult);
+
+        Assert.assertEquals(Lists.newArrayList("DbName", "Type"),
```
Comment (stephen-shelby) [inline fe/fe-core/src/test/java/com/starrocks/common/proc/ExternalProcDirTest.java:87]: showing the type of db make no sense, and the type is catalog type in here.

File: fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalDbsProcDir.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+        // get info
+        List<List<Comparable>> dbInfos = new ArrayList<List<Comparable>>();
+        for (String dbName : dbNames) {
+            Database db = metadataMgr.getDb(catalogName, dbName);
+            if (db == null) {
+                continue;
+            }
+            List<Comparable> dbInfo = new ArrayList<Comparable>();
+            db.readLock();
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalDbsProcDir.java:51]: there is no sense to add db readlock

File: fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTablesProcDir.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Override
+    public ProcResult fetchResult() throws AnalysisException {
+        MetadataMgr metadataMgr = GlobalStateMgr.getCurrentState().getMetadataMgr();
+        Preconditions.checkNotNull(metadataMgr);
+        List<String> tables = null;
+        try {
+            tables = metadataMgr.listTableNames(catalogName, dbName);
+        } catch (DdlException e) {
+            throw new SemanticException(String.format("Get external tables error: ", e.getMessage()));
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTablesProcDir.java:68]:
```suggestion
            throw new SemanticException(String.format("Get external tables error: %s", e.getMessage()));
```

File: fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTablesProcDir.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        // get info
+        List<List<Comparable>> tableInfos = new ArrayList<List<Comparable>>();
+        for (String tableName : tables) {
+            Table table = metadataMgr.getTable(catalogName, dbName, tableName);
+            if (table == null) {
+                continue;
+            }
+            List<Comparable> tableInfo = new ArrayList<Comparable>();
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTablesProcDir.java:77]: getTable will force hive metacache to load table. so if you only want to get type and comment in here, I think it's unnecessary. It's enough to list table names in here.

File: fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTablesProcDir.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        MetadataMgr metadataMgr = GlobalStateMgr.getCurrentState().getMetadataMgr();
+        Database db = metadataMgr.getDb(catalogName, dbName);
+        if (db == null) {
+            throw new AnalysisException("db: " + dbName + " not exists");
+        }
+        db.readLock();
+        Table tbl = null;
+        try {
+            tbl = metadataMgr.getTable(catalogName, dbName, name);
+            if (tbl == null || tbl.isOlapTable()) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTablesProcDir.java:51]: I think the OLAP Table can't get here.

File: fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalSchemaProcNode.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    @Override
+    public ProcResult fetchResult() throws AnalysisException {
+        Preconditions.checkNotNull(table);
+
+        BaseProcResult result = new BaseProcResult();
+        result.setNames(TITLE_NAMES);
+
+        List<Column> schema = table.getFullSchema();
+        List<String> partitionColumns = Lists.newArrayList();
+        if (table.isHiveTable()) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalSchemaProcNode.java:40]:
```suggestion
        if (table.getType != TableType.Hive) {
```

File: fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTableProcDir.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    public static final ImmutableList<String> TITLE_NAMES = new ImmutableList.Builder<String>()
+            .add("Nodes")
+            .build();
+
+    public static final String SCHEMA = "schema";
+    private static final String PARTITIONS = "partitions";
+    private static final ImmutableList<String> CHILDREN_NODES = new ImmutableList.Builder<String>()
+            .add(PARTITIONS)
+            .add(SCHEMA)
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTableProcDir.java:23]: you may add TODO for implementing show partitions in future

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 110 / 157 (70.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/DescribeStmt.java|0|2|00.00%|[132, 133]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|3|7|42.86%|[239, 244, 245, 247]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|11|23|47.83%|[197, 203, 209, 249, 276, 356, 357, 359, 360, 361, 362, 374]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTablesProcDir.java|22|35|62.86%|[35, 40, 41, 43, 44, 45, 46, 48, 49, 50, 52, 62, 63]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTableProcDir.java|16|23|69.57%|[47, 52, 54, 55, 59, 60, 62]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalDbsProcDir.java|24|33|72.73%|[36, 37, 41, 49, 62, 67, 68, 69, 71]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalSchemaProcNode.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/CatalogUtils.java|10|10|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogUtils.java [hunk 11]
```diff
@@ -77,4 +81,19 @@ public static void checkIsLakeTable(String dbName, String tableName) throws Anal
             db.readUnlock();
         }
     }
+
+    public static void convertToMetaResult(BaseProcResult result, List<List<Comparable>> infos) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogUtils.java:85]: you may create a class named "com.starrocks.common.util.ProcResultUtils"

File: fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTableProcDir.java [hunk 12]
```diff
@@ -17,6 +17,8 @@ public class ExternalTableProcDir implements ProcDirInterface {
             .build();
 
     public static final String SCHEMA = "schema";
+
+    // todo implementing show proc external table partitions
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTableProcDir.java:21]: // TODO: implementing show proc external table partitions

File: fe/fe-core/src/main/java/com/starrocks/analysis/DescribeStmt.java (hunk 1 above)
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/analysis/DescribeStmt.java:83]: unuse var?

File: fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTableProcDir.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    public ProcNodeInterface lookup(String entryName) throws AnalysisException {
+        Preconditions.checkNotNull(table);
+
+        if (Strings.isNullOrEmpty(entryName)) {
+            throw new AnalysisException("Entry name is null");
+        }
+
+        // todo add partitions
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/ExternalTableProcDir.java:58]: // TODO: add partitions

File: fe/fe-core/src/main/java/com/starrocks/analysis/DescribeStmt.java (hunk 1 above)
Comment (leoyy0316) [reply fe/fe-core/src/main/java/com/starrocks/analysis/DescribeStmt.java]: no, deleted done

Comment (leoyy0316) [reply fe/fe-core/src/main/java/com/starrocks/analysis/DescribeStmt.java]: yes deleted done

**Comment by stephen-shelby** [comment]
LGTM @HangyuanLiu PTAL, thanks

**Review by HangyuanLiu — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/server/CatalogMgr.java [hunk 14]
```diff
@@ -199,6 +204,10 @@ private static class SerializeData {
 
     }
 
+    public ConcurrentHashMap<String, Catalog> getCatalogs() {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/server/CatalogMgr.java:207]: is it only used for UT, could you mock `createCatalog` in UT instead.

Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/server/CatalogMgr.java:207]: +1, you should delete this function.

Comment (leoyy0316) [reply fe/fe-core/src/main/java/com/starrocks/server/CatalogMgr.java]: done

**Comment by leoyy0316** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 109 / 154 (70.78%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|2|6|33.33%|[235, 240, 241, 243]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|11|22|50.00%|[241, 247, 253, 293, 320, 400, 402, 403, 404, 405, 417]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTablesProcDir.java|22|35|62.86%|[35, 40, 41, 43, 44, 45, 46, 48, 49, 50, 52, 62, 63]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalTableProcDir.java|16|23|69.57%|[47, 52, 54, 55, 59, 60, 62]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalDbsProcDir.java|24|33|72.73%|[36, 37, 41, 49, 62, 67, 68, 69, 71]|
|:large_blue_circle:|com/starrocks/common/util/ProcResultUtils.java|10|11|90.91%|[10]|
|:large_blue_circle:|com/starrocks/common/proc/ExternalSchemaProcNode.java|24|24|100.00%|[]|

**Comment by leoyy0316** [comment]
run starrocks_admit_test

**Review by imay — APPROVED** [review]

**Pull request merged**